File: beagle-ui/src/main/java/br/com/zup/beagleui/framework/engine/renderer/layout/ContainerViewRenderer.kt
package br.com.zup.beagleui.framework.engine.renderer.layout

import android.view.View
import br.com.zup.beagleui.framework.engine.renderer.LayoutViewRenderer
import br.com.zup.beagleui.framework.engine.renderer.RootView
import br.com.zup.beagleui.framework.engine.renderer.ViewRendererFactory
import br.com.zup.beagleui.framework.view.ViewFactory
import br.com.zup.beagleui.framework.widget.core.Flex
import br.com.zup.beagleui.framework.widget.core.FlexDirection
import br.com.zup.beagleui.framework.widget.core.JustifyContent
import br.com.zup.beagleui.framework.widget.layout.Container
import br.com.zup.beagleui.framework.widget.layout.Expanded

internal class ContainerViewRenderer(
    private val container: Container,
    viewRendererFactory: ViewRendererFactory = ViewRendererFactory(),
    viewFactory: ViewFactory = ViewFactory()
) : LayoutViewRenderer(viewRendererFactory, viewFactory) {

    override fun build(rootView: RootView): View {
        val flex = Flex(
            flexDirection = FlexDirection.COLUMN,
            justifyContent = JustifyContent.SPACE_BETWEEN
        )
        val container = viewFactory.makeBeagleFlexView(rootView.getContext(), flex)

        this.container.header?.let {
            container.addView(viewRendererFactory.make(it).build(rootView))
        }

        val expanded = Expanded(this.container.content)
        val contentView = viewRendererFactory.make(expanded).build(rootView)
        container.addView(contentView)

        this.container.footer?.let {
            container.addView(viewRendererFactory.make(it).build(rootView))
        }

        return container
    }
}

File: Sources/BeagleUI/Sources/Engine/Renderer/Renderers/Layout/ContainerViewRenderer.swift
//
//  ContainerViewRenderer.swift
//  BeagleUI
//
//  Created by Daniel Tes on 16/10/19.
//  Copyright © 2019 Daniel Tes. All rights reserved.
//

import UIKit

final class ContainerViewRenderer: WidgetViewRenderer<Container> {
    
    // MARK: - Public Functions
    
    override func buildView() -> UIView {
        let view = UIView(frame: .zero)
        let flex = Flex(flexDirection: .column, justifyContent: .spaceBetween)
        
        if let header = widget.header {
            let headerView = rendenrerProvider.buildRenderer(for: header).buildView()
            view.addSubview(headerView)
        }
        
        let scrollView = contentScrollView()
        view.addSubview(scrollView)
        
        if let footer = widget.footer {
            let footerView = rendenrerProvider.buildRenderer(for: footer).buildView()
            view.addSubview(footerView)
        }
        
        flexViewConfigurator.applyFlex(flex, to: view)
        return view
    }
    
    // MARK: - Private Functions
    
    private func contentScrollView() -> UIScrollView {
        let scrollView = UIScrollView(frame: .zero)
        let flex = Flex(grow: 1)
        
        let contentView = rendenrerProvider.buildRenderer(for: widget.content).buildView()
        scrollView.addSubview(contentView)
        
        flexViewConfigurator.configFlex(flex, to: scrollView)
        return scrollView
    }
}
